docs/source/ProgrammingGuide/Kokkos-and-Virtual-Functions.md

```diff
@@ -1,184 +1,224 @@
 
-## Then why doesn't my code work?
+## Then why the straightforward approach doesn't work?
 
-The reason the intro code might break is that when dealing with GPU-compatible classes with virtual functions, there isn't one Vtable, but two. The first holds the host versions of the virtual functions, while the second holds the device functions. 
+The reason the straightforward approach described above fails is that when dealing with GPU-compatible classes with virtual functions, there isn't one Vtable, but two. The first holds the host version of the virtual functions, while the second holds the device functions.
 
 ![VTableDevice](./figures/VirtualFunctions-VTablesHostDevice.png)
 
-Since we construct the class instance on the host, so it's Vpointer points to the host Vtable.
+Since we construct the class instance on the host, its Vpointer points to the host Vtable.
 
 ![VPointerToHost](./figures/VirtualFunctions-VPointerToHost.png)
 
-Our cudaMemcpy faithfully copied all of the members of the class, including the Vpointer merrily pointing at host functions, which we then call on the device. 
+We faithfully copied all of the members of the class on the GPU memory, including the Vpointer happily pointing at host functions, which we then call on the device.
 
-## How to fix this
+## Make it work
 
-The problem here is that we are constructing the class on the Host. If we were constructing on the Device, we'd get the correct Vpointer, and thus the correct functions (but only for calls on the device). In pseudocode, we want to move from
+The problem here is that we are constructing the instance on the Host.
+If we were constructing it on the Device, we'd get the correct Vpointer, and thus the correct functions.
+Note that this would allow to call virtual functions on the device only, not on the host anymore.
 
-```c++
-Base* hostInstance = new Derived(); // allocate and initialize host
-Base* deviceInstance; // cudaMalloc'd to allocate
-cudaMemcpy(deviceInstance, hostInstance); // to initialize the deivce
-Kokkos::parallel_for(... {
-  // use deviceInstance
-});
-```
+To that aim, we first allocate memory on the device, then construct on the device using a technique called *placement new*
 
-To one where we construct on the device using a technique called `placement new`
+```cpp
+#include <Kokkos_Core.hpp>
 
-```c++
-Base* deviceInstance; // cudaMalloc'd to allocate it
-Kokkos::parallel_for(... {
-  new((Derived*)deviceInstance) Derived(); // construct an instance in the place, the pointer deviceInstance points to
-});
+class Base {
+  public:
+  void Foo() {}
+
+  KOKKOS_FUNCTION
+  virtual void Bar() {}
+};
+
+class Derived : public Base {
+  public:
+  KOKKOS_FUNCTION
+  void Bar() override {}
+};
+
+int main(int argc, char *argv[])
+{
+  Kokkos::ScopeGuard kokkos(argc, argv);
+
+  // create
+  auto deviceInstanceMemory = Kokkos::kokkos_malloc(sizeof(Derived)); // allocate memory on device
```

Comment:
```suggestion
  void* deviceInstanceMemory = Kokkos::kokkos_malloc(sizeof(Derived)); // allocate memory on device
```

Reply:
I was wondering if it was better to keep it `auto`, because `void*` is kind of a joker type.

Reply:
Well it is the return type of `kokkos_malloc`

Reply:
Ok, fair.